### PR TITLE
[clickhouse-cpp] Update to 2.2.1

### PR DESCRIPTION
**Pull request by Thomas1664**
**Describe the pull request**

- #### What does your PR fix?
  Update to 2.2.1
  Use dependencies via vcpkg
  Add feature openssl
  Don't build static and shared libs at the same time

### CODE REVIEW

File: ports/clickhouse-cpp/portfile.cmake [hunk 1]
```diff
@@ -1,18 +1,25 @@
-vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+if(VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+endif()
```
Comment (Thomas1664) [inline ports/clickhouse-cpp/portfile.cmake:3]: Abseil does not dll_export `class absl::int128 __cdecl absl::operator/(class absl::int128,class absl::int128)`:

```
LINK Pass 1: command "C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1433~1.316\bin\Hostx64\x64\link.exe clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\base\compressed.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\base\input.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\base\output.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\base\platform.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\base\socket.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\base\wire_format.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\array.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\column.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\date.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\decimal.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\enum.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\factory.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\ip4.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\ip6.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\lowcardinality.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\nullable.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\numeric.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\string.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\tuple.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\uuid.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\columns\itemview.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\types\type_parser.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\types\types.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\block.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\client.cpp.obj clickhouse\CMakeFiles\clickhouse-cpp-lib.dir\query.cpp.obj /out:clickhouse\clickhouse-cpp-lib.dll /implib:clickhouse\clickhouse-cpp-lib.lib /pdb:clickhouse\clickhouse-cpp-lib.pdb /dll /version:0.0 /machine:x64 /nologo /debug /INCREMENTAL D:\installed\x64-windows\debug\lib\cityhash.lib D:\installed\x64-windows\debug\lib\lz4d.lib wsock32.lib ws2_32.lib D:\installed\x64-windows\debug\lib\abseil_dll.lib -ignore:4221 kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:clickhouse\CMakeFiles\clickhouse-cpp-lib.dir/intermediate.manifest clickhouse\CMakeFiles\clickhouse-cpp-lib.dir/manifest.res" failed (exit code 1120) with the following output:
decimal.cpp.obj : error LNK2019: unresolved external symbol "class absl::int128 __cdecl absl::operator/(class absl::int128,class absl::int128)" (??Kabsl@@YA?AVint128@0@V10@0@Z) referenced in function "bool __cdecl `anonymous namespace'::mulOverflow<int>(class absl::int128 const &,int const &,class absl::int128 *)" (??$mulOverflow@H@?A0xd80ac12a@@YA_NAEBVint128@absl@@AEBHPEAV12@@Z)
clickhouse\clickhouse-cpp-lib.dll : fatal error LNK1120: 1 unresolved externals
```

File: ports/clickhouse-cpp/vcpkg.json [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
   "dependencies": [
+    "abseil",
     "cityhash",
     "lz4",
     {
       "name": "vcpkg-cmake",
       "host": true
-    },
-    {
-      "name": "vcpkg-cmake-config",
```
Comment (Thomas1664) [inline ports/clickhouse-cpp/vcpkg.json:14]: Port doesn't provide CMake config files

File: ports/clickhouse-cpp/portfile.cmake [hunk 3]
```diff
@@ -1,18 +1,26 @@
...
-    SHA512 222b31b16744af64f0a874ec956568adcecb553e43f8d4a2d16c00d55b31015d917a4dc7bb30d5430a894459b1be5e05b292e2d0918bf6f5609046a60539f80f
+    REF v2.2.1
+    SHA512 cf9f14e6bdbf2b739a25004c8f30ef8057cf4afa618c91fc2672059869cbbbdafb72f3027863b3f731f7f2cc239d5690e5e87301bf7930b79fe71d7a4ae3f833
     HEAD_REF master
-    PATCHES 
-        fix-error-c2668.patch
-        fix-error-C4996.patch  #fix x64-uwp error:std::uncaught_exception() is deprecated in C++17
+    PATCHES
+        fix-deps-and-build-type.patch
+        fix-uwp.patch
```
Comment (Thomas1664) [inline ports/clickhouse-cpp/portfile.cmake:13]: Fixes

```
D:\buildtrees\clickhouse-cpp\src\v2.2.1-26f006ec81.clean\clickhouse\base\socket.cpp(30): error C2664: 'DWORD FormatMessageW(DWORD,LPCVOID,DWORD,DWORD,LPWSTR,DWORD,va_list *)': cannot convert argument 5 from 'char [255]' to 'LPWSTR'
D:\buildtrees\clickhouse-cpp\src\v2.2.1-26f006ec81.clean\clickhouse\base\socket.cpp(30): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or parenthesized function-style cast
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winbase.h(2476): note: see declaration of 'FormatMessageW'
```

File: ports/clickhouse-cpp/fix-error-C4996.patch [hunk 4]
```diff
@@ -1,15 +0,0 @@
...
-index 7e10ffd..cbbffdc 100644
---- a/clickhouse/CMakeLists.txt
-+++ b/clickhouse/CMakeLists.txt
-@@ -31,6 +31,10 @@ SET ( clickhouse-cpp-lib-src
-     query.cpp
- )
- 
-+if (MSVC)
-+    add_compile_options(/wd4996)
-+endif()
```
Comment (Thomas1664) [inline ports/clickhouse-cpp/fix-error-C4996.patch:11]: No C4996 generated anymore

File: ports/clickhouse-cpp/fix-error-c2668.patch [hunk 5]
```diff
@@ -1,13 +0,0 @@
...
-+++ b/clickhouse/columns/decimal.cpp
-@@ -27,7 +27,7 @@ inline bool mulOverflow(const Int128 & l, const T & r, Int128 * result)
- 
- #else
- template <typename T>
--inline bool getSignBit(const T & v)
-+inline bool getSignBit(const T & (long double v))
- {
-     return std::signbit(v);
- }
```
Comment (Thomas1664) [inline ports/clickhouse-cpp/fix-error-c2668.patch:13]: Was refactored: https://github.com/ClickHouse/clickhouse-cpp/blob/master/clickhouse/columns/decimal.cpp#L30

File: ports/clickhouse-cpp/portfile.cmake (hunk 3 above)
Comment (LilyWangLL) [reply ports/clickhouse-cpp/portfile.cmake:13]: Does this error fixed by upstream?

Comment (Thomas1664) [reply ports/clickhouse-cpp/portfile.cmake:13]: This error exists at upstream's master as well: https://github.com/ClickHouse/clickhouse-cpp/blob/master/clickhouse/base/socket.cpp#L30

Comment (LilyWangLL) [reply ports/clickhouse-cpp/portfile.cmake:13]: Could you also submit an issue or PR on upstream to fix this issue?

Comment (Thomas1664) [reply ports/clickhouse-cpp/portfile.cmake:13]: Upstream PR: https://github.com/ClickHouse/clickhouse-cpp/pull/230

**Review by LilyWangLL — APPROVED** [review]

File: ports/clickhouse-cpp/fix-deps-and-build-type.patch [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+     IF (BUILD_BENCHMARK)
+diff --git a/clickhouse/CMakeLists.txt b/clickhouse/CMakeLists.txt
+index d96ff88..ad1cb0d 100644
+--- a/clickhouse/CMakeLists.txt
++++ b/clickhouse/CMakeLists.txt
+@@ -36,19 +36,21 @@ IF (WITH_OPENSSL)
+     LIST(APPEND clickhouse-cpp-lib-src base/sslsocket.cpp)
+ ENDIF ()
+ 
+-ADD_LIBRARY (clickhouse-cpp-lib SHARED ${clickhouse-cpp-lib-src})
```
Comment (BillyONeal) [inline ports/clickhouse-cpp/fix-deps-and-build-type.patch:29]: I was going to push a change that minimized this patch: the indenting of this to be inside the BUILD_SHARED_LIBS was unnecessary. However, the TARGET_LINK_LIBRARIES are also different, so it made sense to leave this alone.

Thanks for your contribution to vcpkg!

**Pull request merged**